thunder/tests/distributed/test_ddp.py

```diff
@@ -199,6 +205,78 @@ def fwd_loss(m, x):
             fwd_loss(model, x)
 
+    @pytest.mark.skipif(torch.cuda.device_count() < 2, reason="Requires 2 devices")
+    def test_ddp_weight_sharing(self):
+        # This test is to verify that weight sharing works with ddp.
+        device = torch.device("cuda", self.rank)
+
+        class Model(torch.nn.Module):
+            def __init__(self) -> None:
+                super().__init__()
+                self.fc1 = torch.nn.Linear(16, 16, bias=False)
+                self.fc2 = torch.nn.Linear(16, 16, bias=False)
+
+            def forward(self, x):
+                return self.fc1(x) + self.fc2(x)
+
+        def _test_model_output_and_gradients(model, x):
+            output = model(x)
+            with device:
+                grad_output = torch.ones_like(output)
+            output.backward(grad_output)
+            expected_shape = (4, 16)
+
+            assert output.shape == expected_shape, f"{output.shape=} - {expected_shape=}"
+
+            # Verify that both params point to same grad tensor.
+            assert id(model.get_parameter("fc1.weight").grad) == id(model.get_parameter("fc2.weight").grad)
+
+            # Verify that we accumulate the gradients for the shared parameter.
+            actual_grad = model.get_parameter("fc1.weight").grad
+            # Based on the forward, grad for both params is `(grad_output.T @ x)`. Multiplying by 2 as the grad will be accumulated.
+            expected_grad = 2 * (grad_output.T @ x)
+            torch.testing.assert_close(actual_grad, expected_grad)
+
+            forward_exec_trace = thunder.last_traces(model)[-1]
+            n_synced_params_forward = 0
+            for bsym in forward_exec_trace.bound_symbols:
+                if bsym.sym.id in (thunder.distributed.prims.PrimIDs.SYNCHRONIZE,):
+                    n_synced_params_forward += 1
+            assert (
+                n_synced_params_forward == 0
+            )  # Assert that no params were synced on forward (they should be removed by later transforms)
+
+            backward_exec_trace = thunder.last_backward_traces(model)[-1]
+            allreduced_grads = 0
+            for bsym in backward_exec_trace.bound_symbols:
+                if bsym.sym.id in (
+                    thunder.distributed.prims.PrimIDs.ALL_REDUCE,
+                    thunder.executors.torchex.all_reduce_prim_impl.id,
+                ):
+                    allreduced_grads += 1
+
+            # The expected behaviour is that the gradients were accumulated (since both weights are the same) and then allreduced, so only one allreduce
+            assert allreduced_grads == 1
+
+        with device:
+            jit_ddp_model = Model()
+            ddp_jit_model = Model()
+            x = torch.ones(4, 16)
+
+        # Check `jit(ddp(model))` works
+        jit_ddp_model.fc1.weight = jit_ddp_model.fc2.weight
+
+        jit_ddp_model = thunder.jit(thunder.distributed.ddp(jit_ddp_model), executors=["torch"])
```

Comment:
out of curiosity, why is `executors=["torch"]`, not the default values which should include nvfuser executor?

Reply:
I based the test on the one from test_fsdp.py. Not sure about why it is that way there, but I went with it. Is it better to leave the default ones?

Reply:
> I based the test on the one from test_fsdp.py. Not sure about why it is that way there, but I went with it. Is it better to leave the default ones?

okay, I wasn't aware of the similar test case for FSDP. It's fine to leave this one as is.

@kshitij12345 would you happen to remember why? It seems that you wrote the reference test case in https://github.com/Lightning-AI/lightning-thunder/pull/490/

Reply:
I think I used the code from issue description for writing the test - https://github.com/Lightning-AI/lightning-thunder/issues/257#issue-2258890286 and forgot to change the executors 😅